editor/tree_search.cpp

```diff
@@ -0,0 +1,481 @@
+			break;
+		}
+	}
+}
+
+void TreeSearchPanel::_bind_methods() {
+	ADD_SIGNAL(MethodInfo("update_requested"));
+	ADD_SIGNAL(MethodInfo("text_submitted"));
+}
+
+TreeSearchPanel::TreeSearchPanel() {
+	this->set_visible(false);
+}
+
+bool TreeSearchPanel::has_focus() {
+	return false;
+}
+
+TreeSearchMode TreeSearchPanel::get_search_mode() {
+	if (!check_button_filter_highlight || !check_button_filter_highlight->is_pressed()) {
+		return TreeSearchMode::HIGHLIGHT;
+	}
+	return TreeSearchMode::FILTER;
+}
+
+String TreeSearchPanel::get_text() {
+	if (!line_edit_search)
+		return String();
+	return line_edit_search->get_text();
+}
+
+void TreeSearchPanel::show_and_focus() {
+	this->set_visible(true);
+	line_edit_search->grab_focus();
+}
+
+/* !TreeSearchPanel */
+
+/* ------- TreeSearch ------- */
+
+void TreeSearch::_filter_tree(const String &p_search_mask) {
+	if (matching_entries.size() == 0) {
+		return;
+	}
+
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		TreeItem *cur_item = ordered_tree_items[i];
+
+		if (number_matches.has(cur_item)) {
+			continue;
+		}
+
+		TreeItem *first_counting_ancestor = cur_item;
+		while (first_counting_ancestor && !number_matches.has(first_counting_ancestor)) {
+			first_counting_ancestor = first_counting_ancestor->get_parent();
+		}
+		if (!first_counting_ancestor || !_vector_has_bsearch(matching_entries, first_counting_ancestor)) {
+			cur_item->set_visible(false);
+		}
+	}
+}
+
+void TreeSearch::_highlight_tree(const String &p_search_mask) {
+	callable_cache.clear();
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		TreeItem *entry = ordered_tree_items[i];
+
+		int num_m = number_matches.has(entry) ? number_matches.get(entry) : 0;
+		if (num_m == 0) {
+			continue;
+		}
+
+		// make sure to also call any draw method already defined.
+		Callable parent_draw_method;
+		if (entry->get_cell_mode(0) == TreeItem::CELL_MODE_CUSTOM) {
+			parent_draw_method = entry->get_custom_draw_callback(0);
+		}
+
+		Callable draw_callback = callable_mp(this, &TreeSearch::_draw_highlight_item).bind(p_search_mask, parent_draw_method);
+
+		// -- this is necessary because of the modularity of this implementation
+		// cache render properties of entry
+		String cached_text = entry->get_text(0);
+		Ref<Texture2D> cached_icon = entry->get_icon(0);
+		int cached_max_width = entry->get_icon_max_width(0);
+		callable_cache[entry] = draw_callback;
+
+		// this removes render properties in entry
+		entry->set_custom_draw_callback(0, draw_callback);
+		entry->set_cell_mode(0, TreeItem::CELL_MODE_CUSTOM);
+
+		// restore render properties
+		entry->set_text(0, cached_text);
+		entry->set_icon(0, cached_icon);
+		entry->set_icon_max_width(0, cached_max_width);
+	}
+}
+
+// custom draw callback for highlighting (bind 2)
+void TreeSearch::_draw_highlight_item(TreeItem *p_tree_item, Rect2 p_rect, String p_search_mask, Callable p_parent_draw_method) {
```

Comment:
Probably better to not bind `search_mask`.

Reply:
[ffe344d](https://github.com/limbonaut/limboai/pull/229/commits/ffe344d16651307c7c44e1f29897c666a516006d) addresses this.